# CSI5386 Natural Language Processing
## Project - Automatic Classification of Poems by Themes
### Aaditya Suri, Ranjan Goyal, Paritosh Pal Singh

In [1]:
#import data
import pandas as pd
import numpy as np

In [2]:
import os
pardir = os.path.abspath(os.path.join(os.getcwd(), os.pardir))
topics = os.path.join(pardir, "Project", "data", "topics")
anger = os.path.join(topics, "anger")
animals = os.path.join(topics, "animal")
children = os.path.join(topics, "children")
father = os.path.join(topics, "father")
friend = os.path.join(topics, "friend")
river = os.path.join(topics, "river")
romance = os.path.join(topics, "romance")
sky = os.path.join(topics, "sky")
water = os.path.join(topics, "water")
winter = os.path.join(topics, "winter")

In [3]:
def fetch_poems(path):
    poems_arr = []
    for p in os.listdir(path):
        poem_path = os.path.join(path, p)
        f = open(poem_path,encoding='utf-8')
        poem_text = ""
        for i in f:
            poem_text += i.strip() + "\t"
        poems_arr.append(poem_text)
        f.close()
    return poems_arr

In [4]:
p_anger = fetch_poems(anger)
p_animal = fetch_poems(animals)
p_children = fetch_poems(children)
p_father = fetch_poems(father)
p_friend = fetch_poems(friend)
p_river = fetch_poems(river)
p_romance = fetch_poems(romance)
p_sky = fetch_poems(sky)
p_water = fetch_poems(water)
p_winter = fetch_poems(winter)

In [5]:
p_anger

["PHYSICAL RESOLVES FOR ANGER:\tClose your eyes,  Focus on breath\tAsk, Am I right? , Affirm, I will not shout\tSpeak in lighter voice (Most  Impossible)\tTake a bath, Listen to favorite music\tGo for a long Walk or a Careful Drive\tClasp palms firmly or\tFold hands across chest\tBe Aware of hands\tLook above your head\tAwareness means Anger is done away\tWalk Left and Right, Walk in circles\tLook at the mirror, Focus on your eyes\tSmile a Bit, Laugh a Bit, Scribble something\tSay, 'You are Bloody Punk, If you get Anger'\tSay, ' I am Godlike, I won't get Anger'\tPlace finger on Mush Area Horizontally or Vertically,\tIt puts physical barrier on Shouting\tMeditate, Practice Yoga, Exercise\tDrink Cold Water, Say 'Cool Down'\tSay 'Om Shanthi' or Amen\tMENTAL RESOLVES FOR ANGER:\tThink Anger is Animalistic, Anger is Fire, Anger is Health Hazard\tAnger is Anti-Relation, Anger is weakness, Think End Result of Anger\tBe willing to overcome Anger, Resolve to end Anger\tThink, Should the Pot (Re

In [6]:
data = []
p_array = [p_anger, p_animal, p_children, p_father, p_friend, p_river, p_romance, p_sky, p_water, p_winter]
data_labels = []
for i in range(0, len(p_array)):
    data.extend(p_array[i])
    data_labels.extend([i]*len(p_array[i]))
data = np.array(data)
data_labels = np.array(data_labels)

In [7]:
print(len(p_anger))
print(len(p_animal))
print(len(p_children))
print(len(p_father))
print(len(p_friend))
print(len(p_river))
print(len(p_romance))
print(len(p_sky))
print(len(p_water))
print(len(p_winter))
print(len(data_labels))

98
100
100
99
100
100
100
100
100
100
997


In [8]:
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('paraphrase-MiniLM-L6-v2')
sentences = p_anger+p_animal+p_children+p_father+p_friend+p_river+p_romance+p_sky+p_water+p_winter
#Sentences are encoded by calling model.encode()
embeddings = model.encode(sentences)
#Print the embeddings
for sentence, embedding in zip(sentences, embeddings):
    #print("Sentence:", sentence)
    #print("Embedding:", embedding)
    print("")

Index order:
- 98: anger
- 100: animal
- 100: children
- 99: father
- 100: friend
- 100:river
- 100:romance
- 100:sky
- 100:water
- 100:winter

    Total : 997

In [9]:
sentences[996]

"When I led by zummer streams\tThe pride o' Lea, as naighbours thought her,\tWhile the zun, wi' evenen beams,\tDid cast our sheades athirt the water;\tWinds a-blowen,\tStreams a-flowen,\tSkies a-glowen,\tTokens ov my jay zoo fleeten,\tHeightened it, that happy meeten.\tThen, when maid an' man took pleaces,\tGay in winter's Chris'mas dances,\tShowen in their merry feaces\tKindly smiles an' glisnen glances;\tStars a-winken,\tDay a-shrinken,\tSheades a-zinken,\tBrought anew the happy meeten,\tThat did meake the night too fleeten.\t"

In [10]:
embeddings[996]

array([-1.17485002e-01,  2.18525425e-01,  2.04604506e-01, -5.68921715e-02,
       -2.30186880e-01,  1.30848452e-01,  6.12883747e-01, -1.15391627e-01,
       -1.11299433e-01,  2.51253247e-02, -1.51670069e-01, -1.30314484e-01,
        1.22535706e-01, -1.32448301e-01, -3.78008559e-02,  4.47710045e-02,
        3.75257283e-02,  2.37163603e-01, -1.32957727e-01, -3.32535133e-02,
       -3.12550515e-01, -4.47282121e-02,  1.80475548e-01, -2.03529820e-01,
       -3.16611260e-01, -1.79226771e-01, -9.94291604e-02,  1.59082711e-01,
        8.91862661e-02,  1.06808141e-01,  2.88803309e-01,  3.97132516e-01,
        1.03469424e-01,  1.71287328e-01, -1.91266194e-01,  2.22024173e-01,
       -1.27104968e-01, -3.54634374e-01, -2.02227712e-01,  1.98774189e-01,
        1.14546508e-01, -1.12452939e-01,  1.52317621e-02,  7.81960040e-02,
       -1.07906982e-01, -7.19112754e-02,  1.15956016e-01, -3.84939462e-03,
       -1.72918215e-01,  1.59342401e-02, -2.02196166e-01, -1.52865946e-01,
       -2.91880630e-02, -

In [17]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(np.array(sentences), data_labels, test_size = 0.2)

In [18]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((797,), (200,), (797,), (200,))

In [19]:
# run this in conda terminal
# conda install -c conda-forge spacy
# python -m spacy download en_core_web_sm

In [20]:
def vectors(nlp, data):
    vectors = []
    for row in data:
        row = str(row)
        doc = nlp(row)
        vectors.append(doc.vector)
    return np.array(vectors)

In [21]:
import spacy
nlp = spacy.load('en_core_web_md')
train_vectors = vectors(nlp, X_train)

In [24]:
test_vectors = vectors(nlp, X_test)

In [25]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, confusion_matrix
def evaluate(y_true, y_pred):
    # print evaluation results for model
    # print confusion matrix
    print("Confusion Matrix:")
    print(confusion_matrix(y_true, y_pred))
    # accuracy
    print("Accuracy: ", accuracy_score(y_true, y_pred))
    # precision
    print("Precision: ", precision_score(y_true, y_pred, average = 'weighted'))
    # recall
    print("Recall ", recall_score(y_true, y_pred, average = 'weighted'))
    # f-measure
    print("F1: ", f1_score(y_true, y_pred, average = 'weighted'))
    print("F1: ", classmet(y_true, y_pred))

## Decision Tree Classifier

In [26]:
from sklearn.tree import DecisionTreeClassifier
dt_model = DecisionTreeClassifier()
dt_model.fit(train_vectors, y_train)
y_pred_dt = dt_model.predict(test_vectors)

In [27]:
evaluate(y_test, y_pred_dt)

Confusion Matrix:
[[9 4 1 3 3 2 0 1 0 0]
 [0 9 3 0 2 1 2 0 0 2]
 [1 0 5 2 1 1 3 2 1 2]
 [0 1 1 4 3 0 1 3 0 2]
 [1 2 2 4 5 0 5 0 2 1]
 [0 1 0 0 1 8 0 1 5 1]
 [2 4 0 2 1 2 3 4 0 4]
 [0 1 0 3 3 3 4 8 1 1]
 [1 0 3 1 1 1 1 4 8 2]
 [0 1 2 1 0 2 0 2 1 9]]
Accuracy:  0.34
Precision:  0.35248038264886433
Recall  0.34
F1:  0.3403172704810335


## Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators = 200)
rf_model.fit(train_vectors, y_train)
y_pred_rf = rf_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_rf)

## SVM Classifier

In [ ]:
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(train_vectors, y_train)
y_pred_svm = svm_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_svm)

## XGBoost Classifier

In [ ]:
from xgboost import XGBClassifier
xg_model = XGBClassifier()
xg_model.fit(train_vectors, y_train)
y_pred_xg = xg_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_xg)

## CatBoost Classifier

In [ ]:
from catboost import CatBoostClassifier
cb_model = CatBoostClassifier()
cb_model.fit(train_vectors, y_train)
y_pred_cb = cb_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_cb)

In [ ]:
## Taking only 3 labels instead

In [ ]:
data = []
p_array = [p_anger, p_sky, p_water]
data_labels = []
for i in range(0, len(p_array)):
    data.extend(p_array[i])
    data_labels.extend([i]*len(p_array[i]))
data = np.array(data)
data_labels = np.array(data_labels)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data, data_labels, test_size = 0.2)

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

In [ ]:
train_vectors = vectors(nlp, X_train)
test_vectors = vectors(nlp, X_test)

## Decision Tree Classifier

In [ ]:
from sklearn.tree import DecisionTreeClassifier
dt_model = DecisionTreeClassifier()
dt_model.fit(train_vectors, y_train)
y_pred_dt = dt_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_dt)

## Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators = 200)
rf_model.fit(train_vectors, y_train)
y_pred_rf = rf_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_rf)

## SVM Classifier

In [ ]:
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(train_vectors, y_train)
y_pred_svm = svm_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_svm)

## XGBoost Classifier

In [ ]:
from xgboost import XGBClassifier
xg_model = XGBClassifier()
xg_model.fit(train_vectors, y_train)
y_pred_xg = xg_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_xg)

## CatBoost Classifier

In [ ]:
cb_model = CatBoostClassifier()
cb_model.fit(train_vectors, y_train)
y_pred_cb = cb_model.predict(test_vectors)

In [ ]:
evaluate(y_test, y_pred_cb)